### ISSUE TITLE
tycho-compiler-plugin can't compile test code (assertJ missing)

### ISSUE THREAD

**Issue by georgebax** [eclipse-tycho/tycho#2468]
Hi, I have the following issue after upgrading Tycho from 2.3.0 to 2.7.5. Following the Release Notes, I marked the project in question with JUnitClassPathContainer, so now the JUnit dependencies are found, but I also need AssertJ, which isn't found in the classPath of the testCompile phase by the tycho-compiler-plugin. 
I also tried using the `extraClasspathElements` configuration of the tycho-compiler-plugin like so:

```
                <plugin>
                	<groupId>org.eclipse.tycho</groupId>
                	<artifactId>tycho-compiler-plugin</artifactId>
                	<version>${tycho-version}</version>
                	<configuration>
                		<extraClasspathElements>
					        <dependency>
					            <groupId>com.tngtech.archunit</groupId>
					            <artifactId>archunit-junit5-api</artifactId>
					            <version>${archunit-version}</version>
					        </dependency>
					        <dependency>
					            <groupId>com.tngtech.archunit</groupId>
					            <artifactId>archunit-junit5-engine</artifactId>
					            <version>${archunit-version}</version>
					        </dependency>
					        <dependency>
					            <groupId>org.mockito</groupId>
					            <artifactId>mockito-core</artifactId>
					            <version>${mockito-version}</version>
					        </dependency>
					        <dependency>
					            <groupId>org.hamcrest</groupId>
					            <artifactId>hamcrest-core</artifactId>
					            <version>${hamcrest-version}</version>
					        </dependency>
					        <dependency>
					            <groupId>org.hamcrest</groupId>
					            <artifactId>hamcrest-library</artifactId>
					            <version>${hamcrest-version}</version>
					        </dependency>
					        <dependency>
					            <groupId>org.assertj</groupId>
					            <artifactId>assertj-core</artifactId>
					            <version>${org.assertj}</version>
					        </dependency>
                		</extraClasspathElements>
                	</configuration>
                </plugin>
``` 

in the <pluginManagement> section. The extended debug output also shows that they have been (probably) picked up, but nothing changes : 
`extraClasspathElements = [Dependency {groupId=com.tngtech.archunit, artifactId=archunit-junit5-api, version=1.0.1, type=jar}, Dependency {groupId=com.tngtech.archunit, artifactId=archunit-junit5-engine, version=1.0.1, type=jar}, Dependency {groupId=org.mockito, artifactId=mockito-core, version=5.1.1, type=jar}, Dependency {groupId=org.hamcrest, artifactId=hamcrest-core, version=2.2, type=jar}, Dependency {groupId=org.hamcrest, artifactId=hamcrest-library, version=2.2, type=jar}, Dependency {groupId=org.assertj, artifactId=assertj-core, version=3.24.2, type=jar}]`

The end of the error : 
```
[ERROR] The method assertThat(double) is undefined for the type MyTest
[ERROR] 202 problems (202 errors)
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.5:testCompile (default-testCompile) on project com.myproject.shared: Compilation failure
```

Do you have any pointers about this?

**Comment by laeubi** [comment]
Can you provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue? In general you should not use `extraClasspathElements` but depending on your configuration maybe an extra requirement.

**Comment by georgebax** [comment]
Hi @laeubi, thanks for the reply. I'll try to get around to creating an integration-test for that, sure. If I don't use the `extraClasspathElements`, though, how can I get the plugin to recognize any test dependencies (that aren't JUnit)? It looks like something that should be easily possible and relate to a standard project, but I didn't find anything in the docs.

**Comment by laeubi** [comment]
You can specify additional test dependencies here:

https://tycho.eclipseprojects.io/doc/latest/tycho-surefire-plugin/test-mojo.html#dependencies

**Comment by georgebax** [comment]
But this is for the test runtime surefire plugin, what about the testCompile phase of the tycho-compiler-plugin ?

**Comment by laeubi** [comment]
There are two cases:

1. you can add an import-package for your compile scoped item to the manifest of the test
2. you can define an extra requirement: https://tycho.eclipseprojects.io/doc/latest/target-platform-configuration/target-platform-configuration-mojo.html#dependency-resolution

it depends a bit on how the project setup is, but usually compile dependencies are also used in the IDE and are handled there already, so how do you consume assertJ there?

**Comment by georgebax** [comment]
Well, I declare the assertJ dependency in the pom.xml as such : 
```
		<dependency>
		    <groupId>org.assertj</groupId>
		    <artifactId>assertj-core</artifactId>
		    <version>${org.assertj}</version>
		    <scope>test</scope>
		</dependency>
```
and it also appears in the mvn dependency:tree as `[INFO] +- org.assertj:assertj-core:jar:3.24.2:test`, and I can also run the tests in the IDE, so it's the compiler plugin that doesn't recognize it. 

As far as I know, there is no way to define test dependencies in the Manifest, and wouldn't want to include an Import-Package statement to assertJ packages, is this what you meant with your first point?

**Comment by laeubi** [comment]
You can enable to look at pom dependencies as well with:

https://tycho.eclipseprojects.io/doc/latest/target-platform-configuration/target-platform-configuration-mojo.html#pomDependencies

**Comment by georgebax** [comment]
I tried this and it isn't working. I've added this to the pluginManagement of the root project and in the debug output it looks like it's picked up by all modules, however I don't see how this is affecting the test-compile phase - it's stated that the `target-platform-configuration` mojo isn't executable, it only holds the configuration, but for whom?

**Comment by georgebax** [comment]
Hi @laeubi , I have created a minimal example of why it looks like this is failing, using now tycho 3.0.4 and jdk 17. I basically have a project containing just a Main class and a unit test with no other dependencies other than junit (version 5.9.1). The pom.xml below (without the commented out repository) uses the platform repositories that we provide through artifacts on a TC build, and these platform repositories pull libraries from the eclipse orbit repository (https://download.eclipse.org/tools/orbit/downloads/drops/R20230302014618/repository/ for example), which **doesn't seem to include junit libraries in version 5.9.2**, only 5.9.1. 
If I enable the eclipse releases/latest repo, the build succeeds, but I wouldn't like to include it, because I can't really see the bunde contents of this repository, only the features.
So it looks like tycho requires junit version 5.9.2, but this doesn't exist in the eclipse repositories somehow - is my thinking correct ? 

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>group</groupId>
	<artifactId>TestProject</artifactId>
	<version>1.0.0-SNAPSHOT</version>
	<packaging>eclipse-plugin</packaging>
	<name>TestProject</name>
	
	<properties>
        <tycho-version>3.0.4</tycho-version>
        <junit-version>5.9.1</junit-version>
        <surefire-plugin-version>3.1.0</surefire-plugin-version>
        <ourproject-tc-build-number>257</ourproject-tc-build-number>
        
        <!-- locations of the P2-repositories for the platform -->
        <platform-aggregated-url>https://teamcity.ourproject.com/guestAuth/repository/download/FDMPipeline_Platform_Build/${ourproject-tc-build-number}/aggregated</platform-aggregated-url>
        <platform-3rdparty-url>https://teamcity.ourproject.com/guestAuth/repository/download/FDMPipeline_Platform_Build/${ourproject-tc-build-number}/3rdparty</platform-3rdparty-url>
	</properties>


<!--    <repositories>-->
<!--        <repository>-->
<!--            <id>maven-central</id>-->
<!--            <name>Central Repository</name>-->
<!--            <url>https://repo.maven.apache.org/maven2</url>-->
<!--            <layout>default</layout>-->
<!--            <snapshots>-->
<!--                <enabled>false</enabled>-->
<!--            </snapshots>-->
<!--        </repository>-->
<!--    </repositories>-->

    <repositories>
        <repository>
            <id>maven-central</id>
            <name>Central Repository</name>
            <url>https://repo.maven.apache.org/maven2</url>
            <layout>default</layout>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </repository>
        <repository>
            <id>platform-aggregated</id>
            <url>${platform-aggregated-url}</url>
            <layout>p2</layout>
        </repository>
        <repository>
            <id>platform-3rdparty</id>
            <url>${platform-3rdparty-url}</url>
            <layout>p2</layout>
        </repository>
		<repository>
			<id>eclipse</id>
			<layout>p2</layout>
			<url>http://download.eclipse.org/releases/latest</url>
		</repository>        
    </repositories>


    <build>
		
        <defaultGoal>verify</defaultGoal>
        <!-- only relevant for modules containing plain java tests -->
        <sourceDirectory>src</sourceDirectory>
        <testSourceDirectory>test</testSourceDirectory>
        
        <plugins>
	        <plugin>
	            <groupId>org.eclipse.tycho</groupId>
	            <artifactId>tycho-maven-plugin</artifactId>
	            <version>${tycho-version}</version>
	            <extensions>true</extensions>
	        </plugin>
            <plugin>
                <groupId>org.eclipse.tycho</groupId>
                <artifactId>target-platform-configuration</artifactId>
                <version>${tycho-version}</version>
                <configuration>
                    <environments>
                        <environment>
                            <os>win32</os>
                            <ws>win32</ws>
                            <arch>x86_64</arch>
                        </environment>
                        <environment>
                            <os>linux</os>
                            <ws>gtk</ws>
                            <arch>x86_64</arch>
                        </environment> 
                    </environments>
                </configuration>
            </plugin>
            
            <plugin>
                <groupId>org.eclipse.tycho</groupId>
                <artifactId>tycho-surefire-plugin</artifactId>
                <version>${tycho-version}</version>
            </plugin>
            
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>${surefire-plugin-version}</version>
                <dependencies>
                    <dependency>
                        <groupId>org.junit.jupiter</groupId>
                        <artifactId>junit-jupiter-engine</artifactId>
                        <version>${junit-version}</version>
                    </dependency>
                </dependencies>
                <executions>
                    <execution>
                        <id>test</id>
                        <phase>test</phase>
                        <configuration>
                            <includes>
                                <include>**/*.java</include>
                            </includes>
                            <excludedGroups>PluginTest</excludedGroups>
                        </configuration>
                        <goals>
                            <goal>test</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>            
            
        </plugins>
	</build>	
	
</project>
```

**Comment by laeubi** [comment]
> Hi @laeubi , I have created a minimal example

Can you share the full example? In general Tycho does not require any junit version, but it might be that your bundle is configured to use this, and gnerally using newer version should not be a problem?

At laest the junit you configure for maven-surefire is independent.

**Comment by georgebax** [comment]
Hi, I have pushed the example to a new repository, please run `./mvnw clean verify` to reproduce : https://github.com/georgebax/testBundle

**Comment by laeubi** [comment]
Thanks for the example, you are using the `JUNIT5` Container here:

https://github.com/georgebax/testBundle/blob/242200034a924214921cf7f3b1b1f1fb8fa25c68/.classpath#L15

this requires currently the Junit Bundles from a recent (I think 2022-09) eclipse release, that is why it works if you add the latest repository.

**Comment by georgebax** [comment]
Ah, I see, thanks for that. I had added the JUnit container in the classpath because it was suggested in the Release Notes for version 2.4.0. I removed it and now the test project is building. I will try these changes on the actual project.

**Comment by georgebax** [comment]
Correction, I removed the JUnit Container entry from the .classpath and it now can't find the Junit dependencies that are defined in the pom.xml. I also tried providing the repositories that contain the junit libraries, and added also the target-platform-configuration plugin with the pomDependencies `consider` option. 
It seems I haven't understood, is this entry in the .classpath actually needed ? Why aren't the JUnit (and assertJ) dendencies not able to be resolved for the testCompile phase ?

**Comment by laeubi** [comment]
`assertj` is not part of the JUnit classpath container so you need to provide it somehow, (e.g. with extrarequirements, pom dependecies, ...).

The JUnit COntainer is used to state you need JUnit without explicitly referencing it, so if it is needed depends on your use-case, it is at laest not strictly required.

As mentioned before, the best is to provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue, this has several advantages:

1. We can inspect and probably adjust the testcase
2. We make sure the error occurs in an isolated manner and is not caused by some other configuration
3. You can easily compare different Tycho versions
4. If something needs to be fixed, it is proved that the fix actually fixes the problem
5. We make sure that in the future nothing breaks such use-case unintentionally.

**Comment by georgebax** [comment]
Hi @laeubi, sorry for the late reply. I actually tried the tycho snapshot build back in June and it solved the problem with the test dependencies to JUnit and assertJ, so I was waiting for the 4.0.0 release. Now that I've made the project use tycho 4.0.1, I've been experiencing a problem (again) in the test-compile phase of a module, even though nothing has been changed in the architecture of the project. 
The module `businesslayer.implementation` is **not** a maven project, and in its `MANIFEST.MF` has dependency to another module named `shared`, which is built before the businesslayer in the build process. The test classes of the `businesslayer.implementation` project can't see _any_ class from the `shared` project, so the error is of the type `import ...shared.util.dto... cannot be resolved`. 
Previous to that, in the tycho debug log I see the following: 
```
[INFO] Compiling 265 source files to /Users/gbax/workenv/....businesslayer.implementation/target/test-classes using Eclipse Compiler for Java(TM) 3.34.0.v20230523-1233
[DEBUG] JDT compiler args: [-properties, /Users/gbax/workenv/...businesslayer.implementation/.settings/org.eclipse.jdt.core.prefs, -s, /Users/gbax/workenv/...businesslayer.implementation/target/generated-sources/annotations, -d, /Users/gbax/workenv/...businesslayer.implementation/target/test-classes, -classpath, 
..........
:+com/.../shared/util/dto/*
..........
```
which tells me that the the required packages are included in the classpath. 

This also happens when I try to build another project (`client`), which also has a test dependency to the shared project, this time clearly defined in the pom.xml.
Note, the source files of these projects are compiled without any problems. 

(**tldr**)
So in short, is there a different requirement/way with tycho 4.X.Y to declare test dependencies ?

I will also try to provide an integration-test for this, but I'm wondering if this is a quick fix that I'm missing.
